lib/matroid.js

```diff
@@ -1,71 +1,332 @@
+    multiple clients with the same token so they don't endlessly refresh each others' tokens
+  */
+  retrieveToken(opts) {
+    return new Promise((resolve, reject) => {
+      let options = opts;
+      if (!options) options = {};
+
+      let { expireToken, requestFromServer } = options;
+
+      if (this.authorizationHeader && !(expireToken || requestFromServer)) {
+        return resolve(this.authorizationHeader);
       }
-    };
-    request(config, (err, res, body) => {
-      let result = safeParse(body);
-      if (result.access_token) {
-        self.authorization_header = `${result.token_type} ${result.access_token}`;
+
+      let requestConfigs = {
+        action: 'token',
+        data: {
+          client_id: this.clientId,
+          client_secret: this.clientSecret,
+          grant_type: 'client_credentials'
+        },
+        noAuth: true
+      };
+
+      if (requestFromServer) {
+        requestConfigs.data.refresh = 'true';
       }
 
-      callback(err, result);
+      this._genericRequest(requestConfigs, this._setAuthToken.bind(this, resolve), reject);
     });
   }
 
-  accountInfo(callback) {
-    if (!callback) {
-      throw new Error('Please provide a callback function');
-    }
-    let { method, uri } = this.endpoints.accountInfo;
+  accountInfo() {
+    return new Promise((resolve, reject) => {
+      let options = {
+        action: 'accountInfo'
+      };
+
+      this._genericRequest(options, resolve, reject);
+    });
+  }
+
+  createDetector(zipFile, name, detectorType) {
+    return new Promise((resolve, reject) => {
+      if (!this._checkFilePayloadSize(zipFile, 0, this._fileSizeLimits.zip)) {
+        throw new Error(`Individual file size must be under ${this._fileSizeLimits.zip / 1024 / 1024}MB`);
+      }
+
+      let options = {
+        action: 'createDetector',
+        data: { name: name, detector_type: detectorType },
+        filePaths: { file: zipFile }
+      };
+
+      this._genericRequest(options, resolve, reject);
+    });
+  }
+
+  trainDetector(detectorId, name, detectorType) {
+    return new Promise((resolve, reject) => {
+      let options = {
+        action: 'trainDetector',
+        uriParams: { ':key': detectorId },
+        data: { name: name, detector_type: detectorType }
+      };
+
+      this._genericRequest(options, resolve, reject);
+    });
+  }
+
+  detectorInfo(detectorId) {
+    return new Promise((resolve, reject) => {
+
+      let options = {
+        action: 'detectorInfo',
+        uriParams: { ':key': detectorId }
+      };
+
+      this._genericRequest(options, resolve, reject);
+    });
+  }
+
+  /*
+    Expected image format: { url: 'https://www.matroid.com/logo.png'} OR { file: ['/home/user/image.jpg', '/home/user/other_image.png'] }
+  */
+  classifyImage(detectorId, image) {
+    return new Promise((resolve, reject) => {
+      if (!detectorId) {
+        throw new Error('Please pass in the ID of the detector you want to use');
```

Comment:
I think using node's `Error` is sufficient enough, were you planning on something more custom?